### PR TITLE
feat(chart-annotation): agent-drawn chart annotations on MarketView

### PR DESCRIPTION
**Pull request by Chen-zexi**
## Summary

Adds an end-to-end **chart-annotation** capability. The agent can draw on the MarketView price chart — trendlines, horizontal price lines, and price-anchored event markers — and persist them per chart instance. Annotations sync live onto the chart and stream into the chat transcript as inline annotation cards.

MarketView is our **TradingView-style chart**, rendered with TradingView's open-source [Lightweight Charts](https://github.com/tradingview/lightweight-charts) library; annotations are drawn as overlays/primitives on top of that same surface, so the agent's drawings sit natively on the price chart a broader audience already recognizes.

## Annotation types shipped

Eight annotation types (`draw_chart_annotation`), each validated by its own schema:

| Type | What it draws |
|---|---|
| `price_line` | Horizontal line at a price level (support/resistance), styleable solid / dashed / dotted |
| `trendline` | Diagonal line connecting two `(time, price)` points |
| `marker` | Bar marker — `arrowUp` / `arrowDown` / `circle` / `square`, positioned above / below / in bar |
| `vertical_line` | Vertical line at a timestamp (mark an event/time boundary) |
| `rectangle` | Rectangular zone over a time × price region (e.g. consolidation / supply-demand zone) |
| `text` | Free text label anchored at a point |
| `event` | Price-anchored event marker with hover/tap detail (used for news events) |
| `fib_retracement` | Fibonacci retracement levels between two anchor points |

Lifecycle is managed by `manage_chart_annotations` with `list` / `remove` / `clear_all` actions.

## Chart instances & multi-step iteration

A **chart instance is keyed by `workspace_id | TICKER:interval`** — annotations are scoped to a specific symbol *and* interval, on a **per-workspace basis**. This is the unit the agent operates on:

- The agent can **iterate on a single chart across multiple steps** — drawing, refining, and adding annotations over several `draw_chart_annotation` calls — and every draw targets the same instance.
- Each draw result carries the **full cumulative** annotation set (a superset of the prior draw), so state is reconstructable from any single draw without diffing.
- Switching ticker or interval is a different instance with its own annotations; instances persist per workspace and survive reloads.

## What's included

**Backend (tools + storage + API)**
- `draw_chart_annotation` and `manage_chart_annotations` tools (`src/tools/chart_annotation/`) with validated schemas — bounded payloads, `Timeframe` enum, and LLM payload validation.
- `chart_annotations` table via migration `014_add_chart_annotations.py`, with a raw-psycopg data layer (`src/server/database/chart_annotation.py`).
- Server endpoint (`src/server/app/chart_annotations.py`) wired into `setup.py`.
- A `chart-annotation` agent skill (`skills/chart-annotation/SKILL.md`) registered in the skills registry.

**MarketView (TradingView-style live chart)**
- Built on TradingView's OSS Lightweight Charts; agent-driven annotation rendering and chart sync via `useAgentAnnotations`, `useChartAnnotationSync`, `useChartOverlays`, and the `chartAnnotationStore`.
- Geometry + primitive helpers (`annotationGeometry.ts`, `agentAnnotationsPrimitive.ts`) place annotations against price/time coordinates as chart primitives.
- Price-anchored **news event annotations** with hover/tap detail (`AgentEventOverlay`).
- A `MarketChartSurface` extraction so the chat panel and MarketView share one chart surface.

**Chat (inline annotation cards)**
- `InlineChartAnnotationCard` renders a draw as a spotlight preview, with the heavy chart surface deferred until the card is opened.
- Because the agent iterates on one chart across several draws, the card **pins at the first (anchor) draw's position** but is fed the **latest cumulative** set; each later draw folds into the activity timeline as an ordinary tool-call row rather than being deduped away. One card per distinct chart instance.
- Tool headline reads `TICKER · <interval>` (e.g. `NVDA · 1D`); a subtle framer-motion legend fade-in honors `prefers-reduced-motion`.
- The three duplicated timeframe→label maps are consolidated into a single `INTERVAL_LABEL` derived from `chartConstants.INTERVALS`.

## Notable design decisions

- **Instance = `workspace | ticker:interval`.** Scoping to symbol + interval per workspace lets the agent build one chart up over many steps while keeping annotations from bleeding across tickers, intervals, or workspaces.
- **One card per chart, pinned to the first draw.** Keeps the transcript stable as the chart is built step by step — the user sees intermediate draws as rows in order, while the rich card always reflects the final state. Avoids remounting/refetching the chart surface on every draw.
- **Cumulative artifacts.** Each draw carries the full annotation set, so the pinned card renders the latest state without reconciling deltas.
- **Shared interval label map.** Single source (`INTERVAL_LABEL`) for every timeframe badge, with `INTERVAL_LABEL[key] ?? key` fallback so unknown keys degrade to the raw value.

## Testing

- Frontend: typecheck (`tsc --noEmit`) and ESLint clean (no new warnings); Vitest coverage for the pin-to-first card position, multi-chart rendering, interval headline, `chartInstanceKey` edge cases, geometry, store, sync, and the inline card.
- Backend: schema parity/bounds tests and endpoint tests (`tests/unit/tools/test_chart_annotation*.py`, `tests/unit/server/app/test_chart_annotations.py`).

## Notes

- i18n: en-US and zh-CN strings added for the new tool rows.
- Large first PR because the branch is published for the first time — it carries the full feature across backend, MarketView, and chat.

## Summary by CodeRabbit

* **New Features**
  * Added a “chart-annotation” skill with tools to draw and manage workspace-scoped annotations, including persisted event/news overlays.
  * Introduced backend endpoints to list chart annotations by symbol/timeframe and bulk-clear annotations.
  * Updated chat and MarketView with inline preview cards, modal rendering, multi-draw grouping, workspace-scoped syncing, and a “Clear annotations” chart control.
* **Documentation**
  * Added skill documentation covering supported annotation types, draw payloads, and management actions.
* **Localization**
  * Added English/Chinese UI strings and updated annotation-related statuses and tool labels.
* **Tests**
  * Added API, schema-parity, and chat/MarketView rendering tests for chart-annotation behavior.

### CODE REVIEW

**Comment by Chen-zexi** [comment]
## Review disposition

Verified every finding from both automated reviews against current code. Two commits pushed for the valid + pure-win items; the rest are declined or deferred with reasons.

- `4b003510` — backend (CR3, CR1, N5)
- `2e9e3c34` — frontend (CR12, CR9, N6, CR5, CR8, CR11, nit1, nit2)

All green after: backend ruff + 73 chart-annotation tests, frontend typecheck, ESLint (0 new warnings), full vitest (1839 tests).

### ✅ Applied

| Finding | Source | Fix |
|---|---|---|
| `manage_chart_annotations` DB calls unguarded | CR | Wrapped list/remove/clear_all in try/except → controlled error tuple, matching `draw_chart_annotation`. |
| `resolveChartId` unchecked casts | CR | `typeof === 'string'` guards before `makeChartId` on the live SSE payload. |
| Flash-workspace fetch no retry | CR | Bounded retry (cancel-on-unmount) so a transient null doesn't disable Fast-mode scoping for the session. |
| `events` redundant in range-effect deps | claude | Removed; `recompute` reads `eventsRef.current` only, so the unsub/resub + ResizeObserver/rAF churn is gone. |
| `workflow_active` always shows compact copy | CR | Added `chat.offloadBusy` (en + zh); `/offload` now shows offload-specific copy. |
| `activeTimeframe` JSDoc says "Normalized" | CR | Corrected — it's intentionally the raw active interval. |
| GET `timeframe` not whitespace-stripped | CR | `.strip()` on the GET lookup (DELETE already did via `make_chart_id`). |
| GET/DELETE params lack `max_length` | claude | Added `max_length` caps mirroring DB column widths (defense-in-depth — see note below). |
| `_resetForTesting` leaks `liveAddListeners` | CR | Cleared on reset (test isolation). |
| `useStockBars` symbol cache key not trimmed | CR | `.trim().toUpperCase()`. |
| Relative cross-package import | CR | Switched to `@/pages/...` alias. |

### ❌ Declined — verified false positives

- **`setAll()` skips `bumpMutation()`** — `setAll` has zero production callers (test-only); the freshness race is unreachable.
- **`useChartOverlays` stale `series` closure** — the candlestick series is created mount-only and never rebuilt (theme/symbol changes use `applyOptions`/`setData`), so the cleanup closure is always the live series.
- **`useAgentAnnotations` old primitive not detached** — guarded by `primitiveSeriesRef.current !== series`; the series is stable for the chart's life, and `chart.remove()` disposes attached primitives on teardown.
- **`chartInstanceKey` lowercase `chart_id` mismatch** — the backend double-uppercases (`symbol.upper()` + `make_chart_id`'s `.strip().upper()`), so a lowercase `chart_id` artifact can't occur.
- **`strip_pg_nul_str` coverage / `.upper()` order** — `timeframe` is a closed `Literal` and `annotation_id` is generated hex; neither can carry a NUL, and `symbol` is already stripped. `str.upper()` doesn't add/alter NUL, so the strip-vs-upper order is a no-op.
- **`handlePreview` stuck `loading:true`** — the early return only fires when the URL already exists (viewer has what it needs) or `activeWorkspaceId` is null (unreachable — `ChatBody` only renders with a workspace).
- **`remove()` early-returns without `bumpMutation`** — intended: with no local bucket there's no local mutation to protect, and bumping would wrongly suppress a legitimate server snapshot. `clear()` follows the same pattern.
- **`agentAnnotationsPrimitive` x1/x2 null fallbacks** — `time1/time2` are min/max-ordered upstream and the same-side off-screen case is already skipped, so `0`/`width` are the correct clamps for the only remaining (opposite-side straddle) case.

### 🟡 Deferred — real but unreachable today

- **`selectedInterval` not re-synced on `timeframe` prop change** — the surface mounts per-artifact with a stable `timeframe`, so there's no live bug; adding a forced sync risks overriding a user's manual interval. Will revisit if the surface gains a long-lived, prop-changing mount.
- **Existing annotations not re-rendered on in-place property change** — annotations are immutable-by-id (the store has add/remove/clear ops only; an edit is modeled as remove + add with a new id), so the early-`continue` path is never exercised. Will add property-diff rendering if an in-place update op is introduced.

### ℹ️ Note on the GET/DELETE `max_length` finding

The cap was added, but the claimed `StringDataRightTruncation` → 500 does not actually occur: these params are used in `WHERE`-clause equality reads, which don't truncate-and-error on overlong bind values — the worst case is an empty result. The cap is input-bounding hygiene, not a 500 fix.

### Possible product decision (not a code change)

On a `1s` view of a US equity, agent annotations (stored under `…:1day`) key to `…:1s` and are not shown, and Clear stays inert. This is the **documented, intentional** per-interval scoping (`chartAnnotationStore.ts`), with the chip "jump to 1D" as the mitigation. Making a `1s` viewer *see* `1day` annotations is a product change, not a bug fix — leaving as-is unless we decide otherwise.

File: src/tools/chart_annotation/tools.py [hunk 1]
```diff
@@ -0,0 +1,380 @@
...
+        all_items = await add_and_list_annotations(
+            workspace_id, chart_id, symbol_upper, timeframe, stored
+        )
+    except Exception as exc:
+        logger.exception("[chart_annotation] persistence failed")
+        return (
+            f"Could not save annotation (persistence unavailable): {exc}. "
+            "The drawing was NOT applied. Please try again in a moment.",
+            {},
+        )
```
[the remark this answers is not in the document]
Comment (Chen-zexi) [reply src/tools/chart_annotation/tools.py]: Fixed in 7438ecca. Dropped the `{exc}` interpolation; the raw exception now only reaches the log via `logger.exception` and the chat gets a stable generic message. Generalized beyond the flagged draw site — the same leak existed in the `list`/`remove`/`clear` manage handlers (added in a later commit, outside this review's diff range), so all four persistence error paths were patched. The two `f"Error: {exc}"` returns left in place are a controlled `ValueError` from our own workspace-id resolution, not DB internals.

**Comment by Chen-zexi** [comment]
Dispositions (verified each against the code):

**1. DELETE `timeframe` not enum-validated — fixed in 6cdacba1.** Confirmed: the value flows into `make_chart_id` as chart identity, so a typo like `1week` produced a misleading `200 {cleared: 0}`. Constrained the param to the `Timeframe` literal (422 on bad input). Left GET's `timeframe` as a free `str | None` (it's a filter, not identity) and DELETE's `symbol` as a free string (open-ended ticker, no enum to validate against). No test changes needed — existing DELETE tests only pass in-set values.

**2. `flashWorkspaceIdPromise` not cleared on logout — fixed in 6cdacba1.** Mechanism confirmed. One correction to the writeup: MarketView's retry loop does **not** self-heal this — it only retries on a falsy id, and a stale cache resolves to a *truthy* previous-user id, so the retry never fires. The real backstop is the backend 403 + `(workspace_id, chart_id)` keying (so no data leak), but annotations would mis-scope until remount. Added `clearFlashWorkspaceCache()` and wired it into the SIGNED_OUT branch right beside the existing `queryClient.clear()` — same per-user-cache-wipe contract.

**3. `useChartAnnotationSync` → React Query — declining for this PR, tracking as a follow-up.** Not a correctness bug: the manual effect already cancels in-flight requests (AbortController + `cancelled` guard) with test coverage, and the hook returns `void` so loading/error state has no consumer. The blocker on converting is the `sinceSeq` race-guard — it must capture the seq at fetch-start and thread *that* value into the write for *that* response. Under React Query, cache-serve within `staleTime` and the global `refetchOnWindowFocus` can fire success with a stale/bundled seq and resurrect annotations a concurrent `clear_all` just removed — the exact race the guard exists to prevent. Making that safe means capturing the seq inside `queryFn` and carrying it through `data`, plus a full rewrite of the 10 mechanics tests — a fiddly refactor, not the pure win it looks like. Better as its own change.

**Comment by Chen-zexi** [comment]
Dispositions (verified each against the code):

**1. GET `timeframe` asymmetric with DELETE — fixed in f0f973b3.** You're right, and this reverses what I said on the previous review where I argued GET should stay a lenient filter. On reflection: `timeframe` is a *closed enum* that names part of the resource, and the conventional behavior for a closed-enum filter is 422 on an unknown value, not a silently empty list. Constrained GET to `Timeframe | None` (omit = all timeframes). Verified regression-free — the only caller that sends `timeframe` to GET is one backend test passing `"1hour"` (valid); the frontend never sends it.

**2. Tool signatures type `timeframe` as bare `str` — fixed in f0f973b3.** Annotated both `draw_chart_annotation` and `manage_chart_annotations` as `Timeframe` to match their `args_schema`. One correction to the rationale: this does **not** close a runtime gap — Python annotations aren't enforced at call time, so a direct Python call with a bad string would still reach Postgres unchecked; `args_schema` remains the only runtime guard. The change is a static-typing/consistency win (type-checkers + readers now see the real contract), which is worth having regardless.

**3. `useChatMessages.ts` comment contradicts implementation — not reproduced, no change.** The comment is accurate. It lives in the **history-replay** branch (`loadConversationHistory` → artifact handler, which deliberately handles only `todo_update`/`html_widget`/`task` and skips `chart_annotation`). The `applyAnnotationArtifact` call is ~2300 lines away in `createStreamEventProcessor` — the **live SSE** path the comment explicitly points to ("only the live stream applies them"). On reload, `useChartAnnotationSync`'s REST fetch rehydrates the store and the inline card renders from the replayed `tool_call_result.artifact`. The two paths were conflated; the comment documents the intended split correctly.

**Nit: redundant `_ANNOTATION_ADAPTER: TypeAdapter` annotation — fixed in f0f973b3.** Dropped; the parametrized type is inferred from `TypeAdapter(Annotation)`.

**Comment by Chen-zexi** [comment]
Dispositions (verified each against the code with subagents):

**1. Raw `interval` for annotation lookups — fixed in 26e214ab. Good catch, this was a real bug.** Confirmed the mismatch: the agent only draws on `VALID_TIMEFRAMES` and stores under the normalized timeframe, so on the `1s` interval (reachable in custom mode for 1s-capable symbols) the chart read `NVDA:1s` while the agent wrote `NVDA:1day` — annotations invisible, Clear never activated. Scope was slightly wider than the writeup: **five** consumers, not three — `useAnnotationsForView`, `useDisplayCleared`, the `interval` arg to `useAgentAnnotations`, the `clearDisplay` key, and `<AgentEventOverlay timeframe>`. Fixed by deriving `annotationInterval = normalizeTimeframe(interval)` once and using it for all five; the raw `interval` is kept for data fetch / live-tick gating / extended-hours. (Note: `MarketChart` has no component test harness, so this ships covered only by the existing `normalizeTimeframe('1s') -> '1day'` unit test, not a render-level regression test — flagging the gap.)

**2. `setAll()` skips `bumpMutation()` — not a bug, declining.** `setAll()` has **zero production callers** — it's invoked only from three test files (confirmed by grep + full git history); it's neither the server-sync path (`setChartsForSymbol`) nor a live mutation (`add`/`remove`/`clear`, which all bump). The race you describe needs `setAll` to run concurrently with an in-flight GET, which can't happen. The documented guard contract (store.ts comment) intentionally scopes bumping to add/remove/clear. Adding the bump would be worse than a no-op: if someone later wired `setAll` into server-sync, the bump would mark every server write as a local mutation and defeat the guard from the other side. The clean direction, if anything, is deleting the unused `setAll` — leaving that as a possible follow-up rather than touching it here.

**3. `timeframe`/`annotation_id` not NUL-stripped — applied in 26e214ab (as consistency, not a fix).** No reachable NUL path exists (`timeframe` is the `Timeframe` Literal, `annotation_id` is a machine-generated `ann_<hex>`, and the dict copy is already NUL-safe via `SafeJson`), so this is defensive consistency only. Did it properly: `workspace_id` was *also* bare, so wrapped all three string binds, not just the two flagged — otherwise the tuple still wouldn't be uniform.

**4. Old primitive not detached on series replace — declining, no reachable path.** The candlestick series is created exactly once (mount-only effect, `[]` deps) and disposed only on unmount; symbol/interval/theme changes reuse it via `setData`/`applyOptions`, and there's no chart-type swap. So `primitiveSeriesRef.current !== series` only ever fires on the initial attach (when there's no old primitive). The hook comment is accurate. The proposed detach is safe but guards a path that doesn't exist — dead code today. Worth revisiting only if a future in-place series swap (e.g. a bar/line chart-type toggle) is added.

**5. `chart_id` not uppercased in grouping — declining; the proposed fix would introduce a bug.** `.toUpperCase()` on the whole `chart_id` turns `NVDA:1day` into `NVDA:1DAY`, which diverges from `makeChartId`/`storeKey` (uppercase symbol, **lowercase** timeframe) and breaks the existing `chartInstanceKey` test. The stated impact is also off: `chartInstanceKey` output is never compared against store keys (different separator `|` vs `||`; it's only used to group cards within one transcript, so it's self-consistent regardless of casing). And no path produces a lowercase `chart_id` — `make_chart_id` always uppercases the symbol. Leaving as-is; if defense-in-depth is ever wanted, it must uppercase only the symbol segment, not the whole string.

**Comment by Chen-zexi** [comment]
Dispositions (verified each against the code, browser/CSSOM semantics checked):

**1. `_make_annotation_id` discards entropy — applied in a1db8465.** Dropped the `[:16]` slice (now full 122-bit hex). Verified safe: the column is `VARCHAR(64)` (migration 014), so the 36-char id fits with headroom, and no test or type pins the length (the only assertion is `startswith("ann_")`).

**2. `list_charts` LIMIT 2000 is per-query, not per-chart — declining.** Real observation, but the cap is a safety valve that can't bite at realistic scale: it needs 2000+ annotations across a single symbol's timeframes, absurd for hand-drawn agent annotations. The `_MAX_ANNOTATION_ROWS` comment already reads "one chart / **one symbol**", and truncation is logged (workspace + symbol). A per-chart lateral join is disproportionate for an unreachable edge. Happy to add per-chart limiting if a 2000+/symbol workload ever becomes real.

**3. Color CSS-injection — the stated PoC is wrong, but it surfaced a REAL `url()` channel; fixed in a1db8465 with a different fix.** The `red; position: fixed; ...` semicolon PoC does **not** work: React inline-style objects set `node.style.background` via the per-property CSSOM setter (not `cssText`), and the CSSOM rejects a `;`-bearing value wholesale — no sibling-property injection (browser- and jsdom-confirmed). **But** the sink is the `background` *shorthand*, which accepts `url(...)`: `style={{ background: 'url(//evil/x.gif)' }}` renders to a real fetch — an exfil/beacon channel. The proposed regex doesn't close it (protocol-relative `//` and relative `url()` pass the allowlist). Robust fix: switch the four swatch/badge sites from the `background` shorthand to the `backgroundColor` longhand, which rejects `url()`/shorthand content while accepting every legitimate color form (hex3/6/8, rgb/rgba legacy + modern slash, hsl, named). Added a regression test asserting a `url()` color yields an empty `backgroundColor` and a valid color still applies. (Canvas/lightweight-charts color sinks are `fillStyle`/option objects — not CSS, not a vector — left as-is.)

**4. Unbounded `ids` list — applied in a1db8465.** Confirmed no `max_length` on the list or items. Added `max_length=100` on the list and `max_length=64` per id. No regression: a full-hex id is 36 chars, and no real `remove` targets 100+ annotations at once.

**Comment by Chen-zexi** [comment]
Verified all 7 findings. Applied N1/N3/N4 (pure-win cleanups), declined the 3 blockers + N2 as false positives or already-mitigated — evidence below. Pushed in `3742bf90`.

### 🚫 Blocking — all declined

**B1 — unhandled rejection → "React 19 error-boundary crash": false positive.** `getOrFetchFlashWorkspaceId()` wraps its entire body in `try { … } catch { …; return null }` (`flashWorkspace.ts:16-32`), so the promise it returns **resolves to `null` on any error and never rejects**. The `.then()` with no `.catch()` therefore has nothing to catch. Separately, the stated mechanism is incorrect even in general: an unhandled promise rejection fires `window.onunhandledrejection`, it does not propagate to a React error boundary. Adding `.catch()` here would be dead code.

**B2 — `max_length=100` "doesn't bound the list": false positive.** Verified empirically against this code in Pydantic v2:
```
>>> ManageChartAnnotationsArgs(symbol='NVDA', action='remove', ids=['x'+str(i) for i in range(200)])
ValidationError: ids — List should have at most 100 items
>>> ManageChartAnnotationsArgs(symbol='NVDA', action='remove', ids=['z'*200])
ValidationError: ids.0 — String should have at most 64 characters
```
Field-level `max_length` on a `list[T]` field bounds the **list length** (100); the inner `Annotated[str, Field(max_length=64)]` bounds each **element**. Both fire. The suggested `Annotated[list[...], Field(max_length=100)]` rewrite is functionally identical to what's already there.

**B3 — read-back fallback returns partial set: intentional, documented tradeoff.** The `except: return [annotation]` path is deliberate fail-soft, documented in the docstring ("the read-back is best-effort … so a transient read error can't mask a successful write"). The write has already committed (autocommit). `raise` would report a *successful* persist as a *failed* draw to the agent — strictly worse: it either misleads the user or triggers a retry. The partial response also self-heals: the frontend's GET sync (`useChartAnnotationSync`) reconciles to the full set, and the store's freshness guard keeps the live add from being clobbered. Read-back failure on the same connection that just wrote is also extremely rare. Keeping fail-soft.

### 💡 Non-blocking

**N1 — dead `.strip()` — applied.** Removed; `timeframe` is a validated `Timeframe` literal so `.strip()` was a no-op.

**N2 — backend color validator: declined.** The `url()` vector was already closed at the render layer in `a1db8465` (all swatch/badge sites use the `backgroundColor` longhand, which rejects `url(...)`; regression-tested in `AgentEventOverlay.test.tsx`). A backend regex is redundant with that fix and carries real regression risk: agent-drawn colors legitimately include `rgba(...)` (translucent rectangle fills), named colors, etc., and an allowlist that's too tight would reject valid annotations while a loose one adds nothing the longhand doesn't already guarantee.

**N3 — trigger-function dependency comment — applied.** Added a comment noting `update_updated_at_column()` is defined in migration 001 and reused here (as in 003/012).

**N4 — dynamic WHERE comment — applied.** Added a comment that `where` is built only from literal clause strings with all values bound as `%s` params — no injection surface.
